radixdlt-java/src/main/java/com/radixdlt/client/application/RadixApplicationAPI.java

```diff
@@ -833,21 +832,53 @@ public UnsignedAtom buildAtomWithFee(List<ParticleGroup> particleGroups) {
 	 * @return a cold single of an atom mapped from an action
 	 */
 	public Single<UnsignedAtom> buildAtom(Action action) {
-		final Observable<Action> allActions = this.collectActionAndEffects(action).flatMapIterable(l -> l);
-		final Observable<ParticleGroup> statelessParticleGroups = allActions.flatMap(a ->
-			Observable
-				.fromIterable(this.statelessActionToParticleGroupsMappers)
-				.flatMapIterable(mapper -> mapper.mapToParticleGroups(a))
-		);
-		final Observable<List<ParticleGroup>> statefulParticleGroups = this.statefulMappersToParticleGroups(allActions);
+		return buildAtom(Collections.singletonList(action));
+	}
 
-		return Observable.concat(statelessParticleGroups, statefulParticleGroups.flatMapIterable(l -> l))
-			.<List<ParticleGroup>>scanWith(
-					ArrayList::new,
-					(a, b) -> Stream.concat(a.stream(), Stream.of(b)).collect(Collectors.toList())
-			)
-			.lastOrError()
-			.map(this::buildAtomWithFee);
+	/**
+	 * Returns a cold single of an unsigned atom given a user action. Note that this is
+	 * method will always return a unique atom even if given equivalent actions
+	 *
+	 * @param actions actions to build a single atom
+	 * @return a cold single of an atom mapped from an action
+	 */
+	public Single<UnsignedAtom> buildAtom(List<Action> actions) {
```

Comment:
Would be more generally useful as `Iterable<Action>` rather than `List<Action>`.

Iterable can be converted to stream easily: `StreamSupport.stream(actions.spliterator(), false)`.

Will also need to change `stageActions(...)` method too.

Reply:
Actually not so sure here now. `actions` should really be an ordered list, using `Iterable`would encourage Sets and Collections to be used which really shouldn't be done.